### ISSUE TITLE
[ROCm] test_mla_backends surfaces two MLA decode-path issues: ROCM_AITER_MLA no query_len>4 kernel on gfx942 + bf16 chunked-context tolerance

### ISSUE THREAD

**Issue by wjabbour** [vllm-project/vllm#55609]
### Summary

`tests/v1/attention/test_mla_backends.py` currently gives the ROCm MLA prefill
backends near-zero coverage — every `test_backend_correctness` /
`test_chunked_context_backend_correctness` case is skip-marked at collection
because the only ROCm prefill backend in the list (`ROCM_AITER_FA`) reports
unavailable unless `VLLM_ROCM_USE_AITER=1`, which no CI job sets (#53944).

While adding `FLASH_ATTN` to the ROCm prefill list to fix that (PR to follow),
running the now-executable tests on an MI300X (gfx942) surfaced two **decode-path**
issues, neither in the prefill backend being enabled. Filing so the skip / tolerance
markers the PR adds have something to point at.

### 1. `ROCM_AITER_MLA` has no persistent MLA-decode kernel for `query_len > 4` on gfx942

`ROCM_AITER_MLA` is appended to `BACKENDS_TO_TEST` unconditionally on ROCm and
advertises multi-token-query support, so `spec_decode_medium` (query_len 8) runs it.
aiter's `get_heuristic_kernel_mla` has no persistent bf16 kernel for
`gqa_ratio>=16, qseqlen>4` on gfx942 (the table stops at qseqlen 4), so the lookup
fails.

- Until [ROCm/aiter#5297](https://github.com/ROCm/aiter/pull/5297), that failure
  was a raw `std::abort()` — it killed the pytest worker, taking down every other
  test in the shard. ROCm/aiter#5297 (asm_mla ctypes error-bridge migration)
  turns it into a catchable `RuntimeError`; validated on MI300X. The fp8-KV
  column of the same abort is tracked upstream as ROCm/aiter#4752.
- `query_len <= 4` (`spec_decode_small`) works. `fp8` KV works (`ROCM_AITER_MLA`
  doesn't support fp8 KV, so only `TRITON_MLA` runs, and it handles query_len 8).
  Only `spec_decode` + `query_len > 4` + `auto` (bf16) KV is affected — 16 cases.

Real fix options (both AMD-side / larger):
- aiter ships the gfx942 `mla_a16w16_qh32_qseqlen4_gqaratio32_ps` kernel family
  (exists for gfx950; folds `qseqlen 8 / gqa 16` → `qseqlen 4 / gqa 32`,
  capture-safe), or
- `RocmAiterMLA` bounds its advertised spec-decode `query_len_support` to <= 4 on
  gfx942 so it never requests the missing kernel (completes the partial
  `_ASM_PADDED_MAX_PS_QLEN` guard, which today only covers `_decode_num_heads < 16`).

The PR skips the 16 `spec_decode` + `query_len > 4` + bf16 cases on ROCm with a
pointer here.

### 2. `atol`/`rtol` = `1e-2` is below one bf16 ULP for `v_head_dim=256` chunked context

`test_chunked_context_backend_correctness` at `v_head_dim=256` (glm), `auto`
(bf16) KV: one element / 2,883,584 differs — reference `2.03125`, backend `2.0`
(0.03125 = 2 bf16 ULP at magnitude ~2; one bf16 ULP there is 0.0156, already above
`atol=1e-2`). mean diff 7.5e-4, p99 7.8e-3.

Not backend-specific: both decode backends (`TRITON_MLA`, `ROCM_AITER_MLA`) show
the identical diff, so it's the AITER-FA `run_prefill_context_chunk` partial —
its online-softmax merge is a reordered reduction vs the single-pass SDPA
reference, so ~1 ULP of genuine fp32 disagreement is expected, and rounding to
adjacent bf16 grid points near a boundary makes it a 2-ULP output gap.
`FLASH_ATTN` prefill happens to round the other way and passes; `deepseek`
(`v_head_dim=128`, shorter reduction) passes.

The comparison tolerance for a bf16 output vs an independent reference has to
allow >= 1 ULP; `rtol=atol=1e-2` doesn't at magnitude ~2. The PR bumps `rtol` to
`2e-2` for the ROCm `v_head_dim=256` + bf16 case only. A wider fix (the shared
`rtol`/`atol` is arguably a hair tight for bf16 across all backends and both
platforms) would need a full CUDA run and belongs in a separate change. The p99
7.8e-3 / 256-vs-128 asymmetry is worth a look from AMD on whether AITER-FA's
chunk-merge error growth at `v_head_dim=256` is improvable, but it's not a bug
signature.

### Repro

MI300X (gfx942), `rocm/vllm-ci` image, aiter `0.1.21.post1`:

```
VLLM_ROCM_USE_AITER=1 pytest -q tests/v1/attention/test_mla_backends.py \
  -k "backend_correctness and spec_decode_medium"
# with ROCm/aiter#5297 applied: 16 failed (RuntimeError), 32 passed
# without: SIGABRT, 0 collected results
```

**Comment by mustafayildirim** [comment]
## gfx942 (MI325X) verification — and a correction to the fallback premise

Ran the full differential on MI325X (gfx942) against **both** aiter builds, using a clean full-package shadow of upstream/main+this-PR over the nightly image (no file-level mixing):

**Baseline (no fix):**

| aiter | Result |
|---|---|
| 0.1.21.post1 (vllm nightly image pin) | **16 failed — SIGABRT** kills the pytest worker at first `spec_decode_medium` case (exactly the issue's count) |
| 0.1.23.dev88 (main, installed in-pod) | **aborts** — `.FFFFFFFF` `Fatal Python error: Aborted` |

**With commit 1 (guard routes gfx942 qlen>4 bf16 to the non-persistent entry):**

| aiter | Result | kernel dispatched |
|---|---|---|
| 0.1.21.post1 | 8 pass / **8 fail** — wrong output, max diff 0.11–0.16 vs SDPA | `mla_dec_stage1_bf16_a16w16_subQ128_mqa128` (a **Gqa=128** kernel answering a Gqa=16/qseqlen-8 call) |
| main | 8 pass / **8 fail** — wrong output, same magnitude | `mla_a16w16_qh16_m32x4` |

So the abort is fixed, but the issue's assumption that "the non-persistent (gqa 16, qseqlen 8) entry covers it" does **not** hold: the manifest row exists but computes wrong verify output on gfx942 in every aiter build tested. `spec_decode_small` (qlen 2) passes 8/8 on both. aiter's own tests never cover this shape (their decode kernel test covers nhead16 only at mtp=1 and nhead128 at mtp=2; the persistent test states "supported: decode_qlen = 1,2,3,4"), and `mla.py`'s fold path requires persistent mode and skips nhead=16 entirely — so there is no working route.

**Commit 2 therefore adds a hard bound**: non-persistent causal bf16 decode with `max_qo_len > 4` now raises a `ValueError` naming the mitigations instead of emitting garbage logits.

**With both commits:**
- gfx942: 8 pass (qlen≤4), 8 fail — all failures are the clean `ValueError`, zero numerical mismatches
- gfx950 (MI350X): **48/48 pass** (`-k "backend_correctness and spec_decode and ROCM_AITER and 1.0-1.0"`) — the bound never fires, persistent path unchanged; earlier full run (96/96 incl. fp8) also clean
- qlen ≤ 4 unchanged on both arches

The proper perf fix for gfx942 qlen>4 needs a working aiter kernel (persistent fold target or a fixed non-persistent entry) — filed upstream with ROCm/aiter with the dispatch evidence above.
